File: sprint5/By lessons/lesson10code1/lesson10code1.js
function siftUp(heap, index) {
    if (index === 1) {
        return;
    }

    const parentIndex = Math.floor(index / 2);
    if (heap[parentIndex - 1] < heap[index - 1]) {
        [heap[parentIndex - 1], heap[index - 1]] = [heap[index - 1], heap[parentIndex - 1]];
        siftUp(heap, parentIndex);
    }
}

function heapAdd(heap, key) {
    const index = heap.length + 1;
    heap.push(key);
    siftUp(heap, index);
}


const heap = [];
heapAdd(heap, 5);
heapAdd(heap, 8);
heapAdd(heap, 3);
console.log(heap); // Output: [8, 5, 3]

File: sprint5/By lessons/lesson10code1/lesson10code1.cs
using System;
using System.Collections.Generic;

class HeapFunctions
{
    static void HeapAdd(List<int> heap, int key)
    {
        int index = heap.Count + 1;
        heap.Insert(index - 1, key);
        SiftUp(heap, index);
    }

    static void SiftUp(List<int> heap, int index)
    {
        if (index == 1) return;

        int parentIndex = index / 2;
        if (heap[parentIndex - 1] < heap[index - 1])
        {
            int temp = heap[parentIndex - 1];
            heap[parentIndex - 1] = heap[index - 1];
            heap[index - 1] = temp;
            SiftUp(heap, parentIndex);
        }
    }

    static void Main(string[] args)
    {
        List<int> heap = new List<int>();
        HeapAdd(heap, 5);
        HeapAdd(heap, 8);
        HeapAdd(heap, 3);
        Console.WriteLine(string.Join(", ", heap)); // Output: 8, 5, 3
    }
}
